agent/src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala

```diff
@@ -230,10 +231,32 @@ class DelightStreamingConnector(sparkConf: SparkConf) extends Logging {
     * Caused by: com.databricks.spark.util.LimitedOutputStream$LimitExceededException: Exceeded 2097152 bytes (current = 2100278)
     */
   private def serializeEvent(event: SparkListenerEvent): Option[String] = {
+    val legacyMethod =
```

Comment:
Drive-by comment: the method lookup is going to be hundreds of times slower than the actual method invocation.

This code will be much faster if you lift the reflection lookups out of the `def serializeEvent` method and cache the results as private fields in this object.

For example, maybe something like:

```scala
private val oldSparkEventToJsonMethod = Try(JsonProtocol.getClass.getDeclaredMethod("sparkEventToJson", classOf[SparkListenerEvent]).toOption
private val newSparkEventToJsonStringMethod = JsonProtocol.getClass.getDeclaredMethod("sparkEventToJsonString", classOf[SparkListenerEvent])
assert(oldSparkEventToJsonMethod.isDefined || newSparkEventToJsonStringMethod.isDefined, "couldn't find spark event to JSON method")

[...]

private def serializeEvent(event: SparkListenerEvent): Option[String] = {

[...]

  if (oldSparkEventToJsonMethod.isDefined) {
    Some(compact(render(oldSparkEventToJsonMethod.invoke(JsonProtocolProxy.jsonProtocol, event).asInstanceOf[JValue])))
  } else {
    Some(newSparkEventToJsonStringMethod.get
       .invoke(JsonProtocolProxy.jsonProtocol, event)
       .asInstanceOf[String]
    )
  }

[...]
```

Reply:
Hey Josh 👋 
Thanks for the review, as someone who only write Scala to update this agent it's good to have feedback from a more seasoned user of the language
I updated the PR to adapt the scala pseudocode snippet if you want to take a look 😄

Reply:
Hey Josh, thanks for the heads-up about this change in Spark 3.4 and your comment in here :)